app/upgrades/v18/upgrades_test.go

```diff
@@ -49,12 +49,155 @@ type UpgradeTestSuite struct {
-	suite.Run(t, new(UpgradeTestSuite))
+func (s *UpgradeTestSuite) TestUpgrade() {
+	dummyUpgradeHeight := int64(5)
+
+	checkStoreAfterUpgrade := s.SetupTestUnbondingRecords()
+	s.ConfirmUpgradeSucceededs("v18", dummyUpgradeHeight)
+	checkStoreAfterUpgrade()
+}
+
+func (s *UpgradeTestSuite) SetupTestUnbondingRecords() func() {
+	chainId := "sommelier-3"
+	epochNumberBefore := uint64(501)
+	epochNumberAfter := uint64(510)
+	stTokenAmount := sdkmath.NewInt(1_000_000)
+
+	// Unbonding #1 (before prop)
+	redemptionRateUnbonded := sdk.MustNewDecFromStr("1.0256449837573494") // Rate used when unbonding occurred
+	redemptionRateRecords := sdk.MustNewDecFromStr("1.0257449837673494")  // Rate that was on record at the time (+0.0001)
+
+	s.Require().Equal(redemptionRateUnbonded.String(), v18.RedemptionRatesBeforeProp[chainId][epochNumberBefore].String(),
+		"example redemption rate from before prop does not match constants - update the test")
+
+	// Unbonding #2 (after prop)
+	redemptionRateAtPropTime := sdk.MustNewDecFromStr("1.025900883208774724") // Rate at prop time
+	redemptionRateAtUpgradeTime := sdk.MustNewDecFromStr("1.03")              // Rate at upgrade time
+	estimatedRedemptionRate := sdk.MustNewDecFromStr("1.027950441604387362")  // Estimated rate used to update record
+	unknownRedemptionRate := sdk.MustNewDecFromStr("1.029")                   // Rate that was used in unbonding
+
+	s.Require().Equal(redemptionRateAtPropTime.String(), v18.RedemptionRatesAtTimeOfProp[chainId].String(),
+		"example redemption rate from time of prop does not match constants - update the test")
+
+	// Calculate native token in the records before the upgrade
+	initialNativeAmount1 := sdk.NewDecFromInt(stTokenAmount).Mul(redemptionRateRecords).TruncateInt()
+	initialNativeAmount2 := sdk.NewDecFromInt(stTokenAmount).Mul(unknownRedemptionRate).TruncateInt()
+
+	// Calculate expected native amounts after upgrade
+	expectedNativeAmount1 := sdk.NewDecFromInt(stTokenAmount).Mul(redemptionRateUnbonded).TruncateInt()
+	expectedNativeAmount2 := sdk.NewDecFromInt(stTokenAmount).Mul(estimatedRedemptionRate).TruncateInt()
+
+	// Create the host zone with redemption rate at time of upgrade
+	s.App.StakeibcKeeper.SetHostZone(s.Ctx, stakeibctypes.HostZone{
+		ChainId:        chainId,
+		RedemptionRate: redemptionRateAtUpgradeTime,
+	})
+
+	// Create redemption records - one before and one after the prop
+	s.App.RecordsKeeper.SetUserRedemptionRecord(s.Ctx, recordtypes.UserRedemptionRecord{
+		Id:                "A",
+		EpochNumber:       epochNumberBefore,
+		HostZoneId:        chainId,
+		StTokenAmount:     stTokenAmount,
+		NativeTokenAmount: initialNativeAmount1,
+	})
+	s.App.RecordsKeeper.SetUserRedemptionRecord(s.Ctx, recordtypes.UserRedemptionRecord{
+		Id:                "B",
+		EpochNumber:       epochNumberAfter,
+		HostZoneId:        chainId,
+		StTokenAmount:     stTokenAmount,
+		NativeTokenAmount: initialNativeAmount2,
+	})
+
+	// Create epoch unbonding records
```

Comment:
Should we test records in other statuses?

Reply:
was considering that too, but didn't want the test to get too bloated. 

Happy to add it in the morn if you think it makes sense!

Reply:
Maybe just one other status to make sure the record filtering works

Reply:
done